### DISCUSSION TITLE
Chords with no beats specified

### DISCUSSION THREAD

**Discussion by phatmann** [no-chris/chord-mark#497] [General]
When converting ChordPro to ChordMark, every chord is given 4 beats in order to create a readable and useful chord chart. However, anyone looking at the markup would not know that the beats are not specified correctly.

Part of the problem is that a chord with 4 beats is shown with no beats. Thus we cannot distinguish between an "imported" chord and one that takes the whole measure. There are several ways to solve this:

- Make all 4 beat chords use 4 dots. Thus "A" becomes "A....". If a chord as no dots (e.g. "A"), then its beats are unspecified. This certainly solves the issue, but makes the ChordMark notation a bit noisier and harder to author.
- Add a notation to indicate that beats are unspecified. For example, "A?".
- Add an explicit notation to indicate that a chord is 4 beats. For example, "A|" or "A%" or "A-".
- Add a tag at top of ChordMark file to indicate whether unadorned chords are 4 beats or unspecified beats.

In the current notation, chords with the longest duration have the shortest markup. This sometimes makes it slightly hard to "parse" as a human. Not sure if this is important or not in the long run, but it does lead to a slightly confused initial impression. Forcing 4 beat chords to have 4 dots would solve this problem.

**Comment by no-chris** [comment]
Thanks for bringing up this concern!

My initial idea was to solve this at the application level rather than at the ChordMark notation level. After import, the editor could show a dialog moving from chord to chord, allowing to quickly specify the number of beats each chord. That functionality could also be triggered on-demand. That idea is still on the drawing board.

I agree the current state can cause confusion. At the same time, I'm not a big fan of the idea of changing the markup to solve a conversion issue.

>Make all 4 beat chords use 4 dots. Thus "A" becomes "A....". If a chord as no dots (e.g. "A"), then its beats are unspecified. This certainly solves the issue, but makes the ChordMark notation a bit noisier and harder to author.

Yes, simplicity and writing efficiency is a core principle that I'd really like to keep. Human readability of the source file was always a second concern to this.

> Add a notation to indicate that beats are unspecified. For example, "A?".

Well, the default value is sometimes - and in fact, very often - the correct one, and then you'll have to remove all the extras `?`

>Add an explicit notation to indicate that a chord is 4 beats. For example, "A|" or "A%" or "A-".
>Add a tag at top of ChordMark file to indicate whether unadorned chords are 4 beats or unspecified beats.

That would be my favorite option for now, or a dialog like "currently all chords have a duration of exactly 1 bar. This is probably inaccurate, etc.". This could be further re-inforced with clear indications on the import dialog, and also with the functionality mentioned above. Also, a warning could be displayed (and dismissed) if a song does not contain any `.` after any chord symbol or any `%`.

**Reply by phatmann** [reply]
An import dialog is not a sufficient solution, since a user might not feeling like dealing with it at the moment -- especially because ignoring it still produces decent charts. And batch import would make it impractical anyway. 

How about adding a simple comment tag at the top of the markup that says the file was imported from ChordPro? Or only show that comment if the user did not take the time to enter the right beats for the chords.

A warning about a file with no "." or "%" is interesting, but I fear in practice this might produce some false positives.

**Reply by no-chris** [reply]
I did not consider batch conversion indeed, in that case, such a warning could make sense.
I'm not sure it should belong to ChordMark itself though, the application/batch conversion script might be a better host for it

**Comment by phatmann** [comment]
Does ChordMark support comments? The importer could add a comment to every imported file that does not have beat information. A related idea is to support meta tags, and have a tag to indicate that beat information is absent. The renderer could then choose how to handle this. 

Tony

**Reply by no-chris** [reply]
No support yet for comments nor metadata, but that is definitely on the todolist.
